cw/cw305/cw_lite_segmented.py

```diff
@@ -0,0 +1,175 @@
+        num_segments: Number of waves per capture, must be in [``num_segments_min``,
+            ``num_segments_max``].  This number determines the number of waves returned
+            by ``capture_and_transfer_waves``.
+        num_segments_actual: Actual number of waves that will be captured. This number
+            depends on the number of samples per wave. The target must create this many
+            trigger events. Read-only.
+    """
+
+    def __init__(self, num_samples=740):
+        """Inits a CwLiteSegmented.
+
+        Args:
+            ip_addr: IP address of the oscilloscope.
+        """
+        self._scope = cw.scope()
+        self._configure_scope()
+        self.num_segments = 1
+        self.num_samples = num_samples
+        self._print_device_info()
+
+    @property
+    def num_segments_min(self):
+        return 1
+
+    @property
+    def num_segments_max(self):
+        return (self._scope.adc.oa.hwMaxSamples // self._scope.adc.samples) - 1
+
+    @property
+    def num_segments_actual(self):
+        # Must round-up to fill the entire buffer.
+        return round(self._scope.adc.oa.hwMaxSamples / self._scope.adc.samples) + 1
+
+    @property
+    def num_segments(self):
+        return self._num_segments
+
+    @num_segments.setter
+    def num_segments(self, num_segments):
+        if not self.num_segments_min <= num_segments <= self.num_segments_max:
+            raise RuntimeError(
+                f"num_segments must be in [{self.num_segments_min}, {self.num_segments_max}]."
+            )
+        self._num_segments = num_segments
+
+    @property
+    def num_samples_min(self):
+        return 1
+
+    @property
+    def num_samples_max(self):
+        # TODO: is this correct?
+        return self._scope.adc.oa.hwMaxSamples
+
+    @property
+    def num_samples(self):
+        return self._num_samples
+
+    @num_samples.setter
+    def num_samples(self, num_samples):
+        if not self.num_samples_min <= num_samples <= self.num_samples_max:
+            raise RuntimeError(
+                f"num_samples must be in [{self.num_samples_min}, {self.num_samples_max}]."
+            )
+        self._num_samples = num_samples
+        # This should ideally be handled by the chipwhisperer library but setting the
+        # number of samples smaller than 241 results in "received fewer points than
+        # expected" error. This number is further rounded up by chipwhisperer so that
+        # (num_samples-1) is divisible by 3. We get the actual number of samples from
+        # adc.
+        self._scope.adc.samples = max(241, num_samples)
+        # Note: CW-Lite actually returns one less than adc.samples.
+        self._num_samples_actual = self._scope.adc.samples - 1
+        if self.num_segments > self.num_segments_max:
+            print(f"Warning: Adjusting number of segments to {self.num_segments_max}.")
+            self.num_segments = self.num_segments_max
+
+    def _configure_scope(self):
```

Comment:
nit: Maybe as a TODO, but it would be good to extract this information, maybe the class should take in a scope instance? Construction wise I think we are doing this in other places.

Reply:
I agree. There are several ways to address this. I already have a related TODO in `simple_capture_traces_batch.py` to remove scope init from `device.py`. The main differences between the batch and single capture modes are `basic_mode` and `fifo_fill_mode`, which we can handle by adding a base class for both modes of cw-lite.